Sources/StreamChatUI/ChatMessageList/ChatMessageListVC.swift

```diff
@@ -533,3 +606,103 @@ open class ChatMessageListVC:
+        var snapshot = diffableDataSource?.snapshot() ?? NSDiffableDataSourceSnapshot<Int, ChatMessage>()
+
+        var updatedMessages: [ChatMessage] = []
+        var removedMessages: [ChatMessage] = []
+        var insertedMessages: [(ChatMessage, row: Int)] = []
+        var movedMessages: [(from: ChatMessage, to: ChatMessage)] = []
+
+        var hasNewInsertions = false
+
+        changes.forEach { change in
+            switch change {
+            case let .insert(message, indexPath):
+                insertedMessages.append((message, row: indexPath.row))
+                hasNewInsertions = indexPath.row == 0
+            case let .update(message, _):
+                updatedMessages.append(message)
+            case let .remove(message, _):
+                removedMessages.append(message)
+            case let .move(_, fromIndex, toIndex):
+                guard let fromMessage = snapshot.itemIdentifiers[safe: fromIndex.row] else { break }
+                guard let toMessage = snapshot.itemIdentifiers[safe: toIndex.row] else { break }
+                movedMessages.append((from: fromMessage, to: toMessage))
+            }
+        }
+
+        let sortedInsertedMessages = insertedMessages
+            .sorted(by: { $0.row < $1.row })
+            .map(\.0)
+
+        if hasNewInsertions, let currentFirstMessage = snapshot.itemIdentifiers.first {
+            // Insert new messages at the bottom.
+            snapshot.insertItems(sortedInsertedMessages, beforeItem: currentFirstMessage)
+        } else if let currentLastMessage = snapshot.itemIdentifiers.last {
+            // Load new messages at the top.
+            snapshot.insertItems(sortedInsertedMessages, afterItem: currentLastMessage)
+        }
+
+        snapshot.deleteItems(removedMessages)
+        snapshot.reloadItems(updatedMessages)
+
+        movedMessages.forEach {
+            snapshot.moveItem($0.from, afterItem: $0.to)
+            snapshot.reloadItems([$0.from, $0.to])
+        }
+
+        // The reason we call `performWithoutAnimation` and `animatingDifferences: true` at the same time
+        // is because we don't want animations, but on iOS 14 calling `animatingDifferences: false`
+        // is the same as calling `reloadData()`. Info: https://developer.apple.com/videos/play/wwdc2021/10252/?time=158
+        UIView.performWithoutAnimation {
+            diffableDataSource?.apply(snapshot, animatingDifferences: true) { [weak self] in
+                let newestMessage = snapshot.itemIdentifiers.first
+                if hasNewInsertions && newestMessage?.isSentByCurrentUser == true {
+                    self?.listView.scrollToMostRecentMessage()
+                }
+
+                // When new message is inserted, update the previous message to hide the timestamp if needed.
+                if hasNewInsertions, let previousMessage = snapshot.itemIdentifiers[safe: 1] {
```

Comment:
The previous cell also needs to be reloaded when the message is deleted

Reply:
This is fixed now too 👍 ✅ Although I which this was not required by our UX. It adds unnecessary complexity if we had a different design